test/tests/functional/pbs_node_rampdown_cset.py

```diff
@@ -0,0 +1,156 @@
+        self.job1_exec_vnode = \
+            "(%s:ncpus=1:mem=2097152kb)+" % (self.n0,) + \
+            "(%s:ncpus=%d:mem=2097152kb)+" % (self.n2, self.ncpus2) + \
+            "(%s:ncpus=%d:mem=2097152kb)" % (self.n3, self.ncpus2)
+
+        # expected values after release of vnode
+        self.job1_schedsel1 = \
+            "1:ncpus=1:mem=2097152kb+1:ncpus=%d:mem=2097152kb" % (self.ncpus2)
+        self.job1_exec_host1 = \
+            "%s/0+%s/0*%d" % (self.n0, self.n1, self.ncpus2)
+        self.job1_exec_vnode1 = \
+            "(%s:ncpus=1:mem=2097152kb)+" % (self.n0,) + \
+            "(%s:ncpus=%d:mem=2097152kb)" % (self.n2, self.ncpus2)
+
+    def test_release_nodes_on_cpuset_sis(self):
+        """
+        Submit job that will use cpus on two NUMA vnodes on second mom,
+        goes in R state. Use pbs_release_node to release one of the NUMA
+        vnodes. The NUMA vnode and its resources get released. Submit a
+        second job requesting the released NUMA vnode. Job should run and
+        a new cpuset gets created that uses the released cpus.
+        """
+        # Submit a job that uses second mom's two NUMA nodes, in R state
+        a = {'Resource_List.select':
+             'ncpus=1:mem=2gb+ncpus=%d:mem=2gb+ncpus=%d:mem=2gb' %
+             (self.ncpus2, self.ncpus2), 'Resource_List.place': 'vscatter'}
+        j1 = Job(TEST_USER, attrs=a)
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {'job_state': 'R',
+                                 'Resource_List.mem': '6gb',
+                                 'Resource_List.ncpus': 1 + self.ncpus2 * 2,
+                                 'Resource_List.nodect': 3,
+                                 'schedselect': self.job1_schedselect,
+                                 'exec_host': self.job1_exec_host,
+                                 'exec_vnode': self.job1_exec_vnode}, id=jid1)
+
+        # Release on NUMA vnode on sis mom using command pbs_release_nodes
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid1, self.n3]
+        ret = self.server.du.run_cmd(self.server.hostname,
+                                     cmd, runas=TEST_USER)
+        self.assertEqual(ret['rc'], 0)
+
+        self.server.expect(JOB, {'job_state': 'R',
+                                 'Resource_List.ncpus': 1 + self.ncpus2,
+                                 'Resource_List.nodect': 2,
+                                 'schedselect': self.job1_schedsel1,
+                                 'exec_host': self.job1_exec_host1,
+                                 'exec_vnode': self.job1_exec_vnode1}, id=jid1)
+
+        # Submit a second job requesting the released NUMA node, in R state
+        a = {'Resource_List.select': 'ncpus=%d:mem=2gb:vnode=%s' %
+             (self.ncpus2 + 2, self.n3)}
+        j2 = Job(TEST_USER, attrs=a)
+        jid2 = self.server.submit(j2)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2, max_attempts=10)
```

Comment:
Shouldn't you be checking to see that the specific vnodes are being used?
Also, don't put a max_attempts here.

Reply:
I have removed the submission of the second job.  I have changed the test to check that the cpuset cpu resource got freed after releasing the specific vnode requested in the initial job.  It is not guaranteed that the system will reuse the recently freed cgroup cpuset resource in the second job.